// mppx/dist/tempo/server/Charge.js
import { decodeFunctionData, isAddressEqual, parseEventLogs } from 'viem';
import { getTransactionReceipt, sendRawTransaction, sendRawTransactionSync, signTransaction, call as viem_call, } from 'viem/actions';
import { tempo as tempo_chain } from 'viem/chains';
import { Abis, Transaction } from 'viem/tempo';
import { PaymentExpiredError } from '../../Errors.js';
import * as Method from '../../Method.js';
import * as Client from '../../viem/Client.js';
import * as Account from '../internal/account.js';
import * as defaults from '../internal/defaults.js';
import * as FeePayer from '../internal/fee-payer.js';
import * as Selectors from '../internal/selectors.js';
import * as Methods from '../Methods.js';
/**
 * Creates a Tempo charge method intent for usage on the server.
 *
 * @example
 * ```ts
 * import { tempo } from 'mppx/server'
 *
 * const charge = tempo.charge()
 * ```
 */
export function charge(parameters = {}) {
    const { amount, currency = defaults.resolveCurrency(parameters), decimals = defaults.decimals, description, externalId, memo, waitForConfirmation = true, } = parameters;
    const { recipient, feePayer, feePayerUrl } = Account.resolve(parameters);
    const getClient = Client.getResolver({
        chain: { ...tempo_chain, experimental_preconfirmationTime: 500 },
        feePayerUrl,
        getClient: parameters.getClient,
        rpcUrl: defaults.rpcUrl,
    });
    return Method.toServer(Methods.charge, {
        defaults: {
            amount,
            currency,
            decimals,
            description,
            externalId,
            memo,
            recipient,
        },
        // TODO: dedupe `{charge,session}.request`
        async request({ credential, request }) {
            const chainId = await (async () => {
                if (request.chainId)
                    return request.chainId;
                if (parameters.testnet)
                    return defaults.chainId.testnet;
                return (await getClient({})).chain?.id;
            })();
            const client = await (async () => {
                try {
                    return await getClient({ chainId });
                }
                catch {
                    throw new Error(`No client configured with chainId ${chainId}.`);
                }
            })();
            if (client.chain?.id !== chainId)
                throw new Error(`Client not configured with chainId ${chainId}.`);
            const resolvedFeePayer = (() => {
                const account = typeof request.feePayer === 'object' ? request.feePayer : feePayer;
                const requested = request.feePayer !== false && (account ?? feePayer ?? feePayerUrl);
                if (credential)
                    return account;
                if (requested)
                    return true;
                return undefined;
            })();
            return {
                ...request,
                chainId,
                feePayer: resolvedFeePayer,
                memo: request.memo || undefined,
            };
        },
        async verify({ credential, request }) {
            const { challenge } = credential;
            const { chainId, feePayer } = request;
            const client = await getClient({ chainId });
            const { request: challengeRequest } = challenge;
            const { amount, methodDetails } = challengeRequest;
            const expires = challenge.expires;
            const currency = challengeRequest.currency;
            const recipient = challengeRequest.recipient;
            if (expires && new Date(expires) < new Date())
                throw new PaymentExpiredError({ expires });
            const memo = methodDetails?.memo;
            const payload = credential.payload;
            switch (payload.type) {
                case 'hash': {
                    const hash = payload.hash;
                    const receipt = await getTransactionReceipt(client, {
                        hash,
                    });
                    if (memo) {
                        const memoLogs = parseEventLogs({
                            abi: Abis.tip20,
                            eventName: 'TransferWithMemo',
                            logs: receipt.logs,
                        });
                        const match = memoLogs.find((log) => isAddressEqual(log.address, currency) &&
                            isAddressEqual(log.args.to, recipient) &&
                            log.args.amount.toString() === amount &&
                            log.args.memo.toLowerCase() === memo.toLowerCase());
                        if (!match)
                            throw new MismatchError('Payment verification failed: no matching transfer with memo found.', {
                                amount,
                                currency,
                                memo,
                                recipient,
                            });
                    }
                    else {
                        const transferLogs = parseEventLogs({
                            abi: Abis.tip20,
                            eventName: 'Transfer',
                            logs: receipt.logs,
                        });
                        const memoLogs = parseEventLogs({
                            abi: Abis.tip20,
                            eventName: 'TransferWithMemo',
                            logs: receipt.logs,
                        });
                        const match = [...transferLogs, ...memoLogs].find((log) => isAddressEqual(log.address, currency) &&
                            isAddressEqual(log.args.to, recipient) &&
                            log.args.amount.toString() === amount);
                        if (!match)
                            throw new MismatchError('Payment verification failed: no matching transfer found.', {
                                amount,
                                currency,
                                recipient,
                            });
                    }
                    return toReceipt(receipt);
                }
                case 'transaction': {
                    const serializedTransaction = payload.signature;
                    const transaction = Transaction.deserialize(serializedTransaction);
                    const calls = transaction.calls ?? [];
                    const call = calls.find((call) => {
                        if (!call.to || !isAddressEqual(call.to, currency))
                            return false;
                        if (!call.data)
                            return false;
                        const selector = call.data.slice(0, 10);
                        if (memo) {
                            if (selector !== Selectors.transferWithMemo)
                                return false;
                            try {
                                const { args } = decodeFunctionData({ abi: Abis.tip20, data: call.data });
                                const [to, amount_, memo_] = args;
                                return (isAddressEqual(to, recipient) &&
                                    amount_.toString() === amount &&
                                    memo_.toLowerCase() === memo.toLowerCase());
                            }
                            catch {
                                return false;
                            }
                        }
                        if (selector === Selectors.transfer) {
                            try {
                                const { args } = decodeFunctionData({ abi: Abis.tip20, data: call.data });
                                const [to, amount_] = args;
                                return isAddressEqual(to, recipient) && amount_.toString() === amount;
                            }
                            catch {
                                return false;
                            }
                        }
                        if (selector === Selectors.transferWithMemo) {
                            try {
                                const { args } = decodeFunctionData({ abi: Abis.tip20, data: call.data });
                                const [to, amount_] = args;
                                return isAddressEqual(to, recipient) && amount_.toString() === amount;
                            }
                            catch {
                                return false;
                            }
                        }
                        return false;
                    });
                    if (!call)
                        throw new MismatchError('Invalid transaction: no matching payment call found', {
                            amount,
                            currency,
                            recipient,
                        });
                    if ((feePayer || feePayerUrl) && methodDetails?.feePayer !== false)
                        FeePayer.validateCalls(calls, { amount, currency, recipient });
                    const resolvedFeeToken = transaction.feeToken ?? defaults.currency[chainId];
                    const serializedTransaction_final = await (async () => {
                        if (feePayer && methodDetails?.feePayer !== false) {
                            return signTransaction(client, {
                                ...transaction,
                                account: feePayer,
                                feePayer,
                                feeToken: resolvedFeeToken,
                            });
                        }
                        return serializedTransaction;
                    })();
                    if (waitForConfirmation) {
                        const receipt = await sendRawTransactionSync(client, {
                            serializedTransaction: serializedTransaction_final,
                        });
                        return toReceipt(receipt);
                    }
                    else {
                        // Optimistic path: simulate to catch obvious reverts, then broadcast
                        // without waiting for on-chain confirmation. The returned receipt
                        // assumes success — callers opt into this risk via waitForConfirmation: false.
                        await viem_call(client, {
                            ...transaction,
                            account: transaction.from,
                            feeToken: resolvedFeeToken,
                            calls,
                        });
                        const hash = await sendRawTransaction(client, {
                            serializedTransaction: serializedTransaction_final,
                        });
                        return {
                            method: 'tempo',
                            status: 'success',
                            timestamp: new Date().toISOString(),
                            reference: hash,
                        };
                    }
                }
                default:
                    throw new Error(`Unsupported credential type "${payload.type}".`);
            }
        },
    });
}
/** @internal */
function toReceipt(receipt) {
    const { status, transactionHash } = receipt;
    if (status !== 'success') {
        throw new Error(`Transaction reverted: ${transactionHash}`);
    }
    return {
        method: 'tempo',
        status: 'success',
        timestamp: new Date().toISOString(),
        reference: transactionHash,
    };
}
/** @internal */
class MismatchError extends Error {
    constructor(reason, details) {
        super([reason, ...Object.entries(details).map(([k, v]) => `  - ${k}: ${v}`)].join('\n'));
        this.name = 'MismatchError';
    }
}
